upstream/jobs.go

```diff
@@ -7,136 +7,293 @@ import (
-			return 0, fmt.Errorf("failed to update is_pushed on config_analysis: %w", err)
-		}
+	return reconcileTable[models.ComponentRelationship](ctx, config, fetcher, batchSize)
+}
 
-		count += len(analyses)
-	}
+func reconcileConfigComponentRelationship(ctx context.Context, config UpstreamConfig, batchSize int) (int, error) {
+	fetcher := ctx.DB().Select("config_component_relationships.*").
+		Joins("LEFT JOIN components c ON config_component_relationships.component_id = c.id").
+		Joins("LEFT JOIN config_items ci ON config_component_relationships.config_id = ci.id").
+		Where("c.agent_id = ? AND ci.agent_id = ?", uuid.Nil, uuid.Nil).
+		Where("config_component_relationships.is_pushed IS FALSE")
+
+	return reconcileTable[models.ConfigComponentRelationship](ctx, config, fetcher, batchSize)
+}
+
+func reconcileCheckComponentRelationship(ctx context.Context, config UpstreamConfig, batchSize int) (int, error) {
+	fetcher := ctx.DB().Select("check_component_relationships.*").
+		Joins("LEFT JOIN components c ON check_component_relationships.component_id = c.id").
+		Joins("LEFT JOIN canaries ON check_component_relationships.canary_id = canaries.id").
+		Where("c.agent_id = ? AND canaries.agent_id = ?", uuid.Nil, uuid.Nil).
+		Where("check_component_relationships.is_pushed IS FALSE")
+
+	return reconcileTable[models.CheckComponentRelationship](ctx, config, fetcher, batchSize)
+}
+
+func reconcileConfigRelationship(ctx context.Context, config UpstreamConfig, batchSize int) (int, error) {
+	fetcher := ctx.DB().Select("config_relationships.*").
+		Joins("LEFT JOIN config_items ci ON config_relationships.config_id = ci.id").
+		Where("ci.agent_id = ?", uuid.Nil, uuid.Nil).
+		Where("config_relationships.is_pushed IS FALSE")
+
+	return reconcileTable[models.ConfigRelationship](ctx, config, fetcher, batchSize)
+}
+
+func reconcileCheckConfigRelationship(ctx context.Context, config UpstreamConfig, batchSize int) (int, error) {
+	fetcher := ctx.DB().Select("check_config_relationships.*").
+		Joins("LEFT JOIN config_items ci ON check_config_relationships.config_id = ci.id").
+		Where("ci.agent_id = ?", uuid.Nil, uuid.Nil).
+		Where("check_config_relationships.is_pushed IS FALSE")
+
+	return reconcileTable[models.CheckConfigRelationship](ctx, config, fetcher, batchSize)
 }
 
-// SyncArtifacts pushes artifacts that haven't already been pushed to upstream.
-func SyncArtifacts(ctx context.Context, config UpstreamConfig, batchSize int) (int, error) {
+// ReconcileTable pushes all unpushed items in a table to upstream.
+func reconcileTable[T dbTable](ctx context.Context, config UpstreamConfig, fetcher *gorm.DB, batchSize int) (int, error) {
 	client := NewUpstreamClient(config)
-	count := 0
+	var anon T
+	table := anon.TableName()
+
+	if !ctx.Properties().Off("upstream.reconcile.pre-check") {
+		var unpushed float64
+		precheck := fmt.Sprintf(`SELECT reltuples FROM pg_class WHERE relname = '%s_is_pushed_idx'`, table)
```

Comment:
This query should be part of PushAll run the query to return the tables with rows to push, and then only push those tables.

```
select c.relname from pg_index i
join pg_class c on c.oid = i.indrelid
join pg_class ic on i.indexrelid = ic.oid
WHERE  i.indexrelid::regclass::text like '%_is_pushed_idx' and  ic.reltuples > 0
```